### PR TITLE
Introduce restoreWallet function in WalletService

### PR DESCRIPTION
**Pull request by JustynaBroniszewska**

### CODE REVIEW

File: universal-login-sdk/src/api/wallet/RestoringWallet.ts [hunk 1]
```diff
@@ -20,9 +20,21 @@ export class RestoringWallet {
...
+      if (isInvalidPasswordError(e) || isInvlidPrivateKeyError(e, this.contractAddress)) {
+        throw new InvalidPassword();
+      } else {
+        throw Error(e);
+      }
+    }
   }
 }
+
+const isInvlidPrivateKeyError = (e: Error, contractAddress: string) => e.message === `Private key is not contract's ${contractAddress} owner`;
```
Comment (xitronix) [inline universal-login-sdk/src/api/wallet/RestoringWallet.ts:38]: isInvAlid...

File: universal-login-sdk/src/api/wallet/RestoringWallet.ts [hunk 2]
```diff
@@ -20,9 +20,21 @@ export class RestoringWallet {
...
+      const wallet = await Wallet.fromEncryptedJson(JSON.stringify(this.encryptedWallet), password);
+      const deployedWallet = new DeployedWallet(this.contractAddress, this.ensName, wallet.privateKey, this.sdk);
+      ensure(await deployedWallet.keyExist(wallet.address), InvalidPrivateKey, this.contractAddress);
+      return deployedWallet;
+    } catch (e) {
+      if (isInvalidPasswordError(e) || isInvlidPrivateKeyError(e, this.contractAddress)) {
+        throw new InvalidPassword();
+      } else {
+        throw Error(e);
+      }
```
Comment (xitronix) [inline universal-login-sdk/src/api/wallet/RestoringWallet.ts:33]: Why do we do it? Catching error and then trow it again is a bad pattern

File: universal-login-sdk/src/api/wallet/RestoringWallet.ts [hunk 3]
```diff
@@ -20,9 +20,21 @@ export class RestoringWallet {
...
 
   async restore(password: string) {
-    const wallet = await Wallet.fromEncryptedJson(JSON.stringify(this.encryptedWallet), password);
-    const deployedWallet = new DeployedWallet(this.contractAddress, this.ensName, wallet.privateKey, this.sdk);
-    ensure(await deployedWallet.keyExist(wallet.address), InvalidPrivateKey, this.contractAddress);
-    return deployedWallet;
+    try {
+      const wallet = await Wallet.fromEncryptedJson(JSON.stringify(this.encryptedWallet), password);
+      const deployedWallet = new DeployedWallet(this.contractAddress, this.ensName, wallet.privateKey, this.sdk);
+      ensure(await deployedWallet.keyExist(wallet.address), InvalidPrivateKey, this.contractAddress);
```
Comment (xitronix) [inline universal-login-sdk/src/api/wallet/RestoringWallet.ts:26]: InvalidPrivateKey  => InvalidPassword

File: universal-login-sdk/src/api/wallet/RestoringWallet.ts (hunk 2 above)
Comment (xitronix) [reply universal-login-sdk/src/api/wallet/RestoringWallet.ts:33]: Remove if else

File: universal-login-sdk/src/core/utils/errors.ts [hunk 4]
```diff
@@ -25,7 +25,8 @@ type ErrorType =
   'InvalidGasLimit' |
   'InvalidObserverState' |
   'SavingFutureWalletFailed' |
-  'InvalidPrivateKey';
+  'InvalidPrivateKey' |
```
Comment (xitronix) [inline universal-login-sdk/src/core/utils/errors.ts:28]: Remove InvalidPrivateKey

File: universal-login-sdk/src/core/utils/errors.ts [hunk 5]
```diff
@@ -25,7 +25,7 @@ type ErrorType =
   'InvalidGasLimit' |
   'InvalidObserverState' |
   'SavingFutureWalletFailed' |
-  'InvalidPrivateKey';
+  'InvalidPrivateKey' ;
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/src/core/utils/errors.ts:28]:
```suggestion
  'InvalidPrivateKey';
```

**Pull request merged**